model-server/src/test/kotlin/org/modelix/model/server/ReplicatedModelTest.kt

```diff
@@ -0,0 +1,154 @@
+
+        // Step 1.3: add a new child node to get a newer version
+        val twoHelloChildrenVersion = addHelloChild(oneHelloChildVersion, client, defaultBranchReference)
+
+        // Step 2: in a new client, fetch the latest repository data
+        val newClient = createModelClient()
+        // we do not provide an initial version, so we expect to fetch the latest one (twoHelloChildrenVersion)
+        val replicatedModel = ReplicatedModel(newClient, defaultBranchReference)
+        val branch = replicatedModel.start { branch ->
+            // Step 3: check, here we must have the twoHelloChildrenVersion (2 "hello" children)
+            val twoHelloChildren = getHelloChildrenOfRootNode(branch)
+            assertEquals(2, twoHelloChildren.size)
+        }
+
+        // Step 4: wait a bit so replicated model can fetch the new versions from the server (in case there would be any -- there is none by the way)
+        delay(500)
+
+        // Step 5: check, eventually we must have the twoHelloChildrenVersion (2 "hello" children)
+        val twoHelloChildren = getHelloChildrenOfRootNode(branch)
+        assertEquals(2, twoHelloChildren.size)
+    }
+
+    @Test
+    fun startsWithSpecificVersion() = runTest {
+        val client = createModelClient()
+        val repositoryId = RepositoryId(UUID.randomUUID().toString())
+        val defaultBranchReference = repositoryId.getBranchReference()
+
+        // Step 1: prepare repository with two versions beside the initial version
+        // Step 1.1: create an empty repository
+        val initialVersion = client.initRepository(repositoryId)
+
+        // Step 1.2: add a new child node to get a new version
+        val oneHelloChildVersion = addHelloChild(initialVersion, client, defaultBranchReference)
+
+        // Step 1.3: add a new child node to get a newer version
+        val twoHelloChildrenVersion = addHelloChild(oneHelloChildVersion, client, defaultBranchReference)
+
+        // Step 2: in a new client, fetch the oneHelloChildVersion
+        val newClient = createModelClient()
+
+        // Step 2.1: to avoid version was not created by this client exception
+        val oneHelloChildVersionClone = newClient.loadVersion(repositoryId, oneHelloChildVersion.getContentHash(), null)
+
+        // Step 2.2: we provide an initial version, so we expect to fetch it first (oneHelloChildVersion)
+        val replicatedModel = ReplicatedModel(
+            newClient,
+            defaultBranchReference,
+            lastKnownRemoteVersion = oneHelloChildVersionClone as CLVersion,
+        )
+        val branch = replicatedModel.start { branch ->
+            // Step 3: check, here we must have the oneHelloChildVersion, which has 1 "hello" child
+            val oneHelloChild = getHelloChildrenOfRootNode(branch)
+            assertEquals(1, oneHelloChild.size)
+        }
+
+        // Step 4: wait a bit so replicated model can fetch the new versions from the server
+        delay(500)
```

Comment:
Can we get this deterministic without waiting?

Reply:
Good question. Because the point is, the ReplicatedModel must poll the latest version which is running in a Job started on the CoroutineScope.

Maybe with a blocking wait and a timeout. That way we do not wait 500 ms, but shorter if the response arrive quicker. Otherwise we wait untilt timeout.